Improve Test Runner's Minitest integration.

kaspth opened the pull request:
I've tried to improve the Rails new test running system to take better advantage of Minitest's extension system.

The build will break as I haven't ported over the Runner tests yet. But the tests running the tests commands should pass.

@zenspider, was this more inline with your comments on the original pull request?

cc @senny @arthurnn

kaspth committed:
Improve test runner's Minitest integration.

This also adds free mix and matching of directories, files and lines filters.
Like so:

bin/rails test models/post_test.rb test/integration models/person_test.rb:26

You can also mix in a traditional Minitest filter:

bin/rails test test/integration -n /check_it_out/

senny commented:
@kaspth Thank you for working on this :yellow_heart: 

Can we make it work without `Minitest.autorun`? Part of the idea of the original runner was to have a simple interface to run a bunch of tests:

```ruby
Rails::TestRunner.run(["test/controllers", "test/mailers", "test/functional"])
```

When relying on `Minitest.autorun` the API feels unclear and relies on the require side-effects. There is a related issue #17708 and PR #19435.

senny commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -1,14 +1,66 @@
+
+    opts.on("-n", "--name [FILENAME:LINE]",
+            "Run a single test by appending the line number to filename:\n\n" \
+            "bin/rails test test/models/user_test.rb:27") do |name|
```
I'm a bit confused by this option. Isn't `-n` supposed to name the method to run? Also the example does not use the flag at all. This seems to be more an explanation of the `[file or directory]` argument at the top.

arthurnn commented:
I still think we should have a `TestRunner` class. maybe that class wont parse the options anymore. But it would be good to have the test runner logic in one class.

kaspth commented:
Alright, updated here @senny @arthurnn

senny commented on railties/test/application/test_runner_test.rb:
```diff
@@ -14,7 +15,7 @@ def teardown
     end
 
-    def test_run_in_test_environment
+    def test_run_in_differing_environments
```
@kaspth do you think it's worth keep the other test around to be sure it runs in test env by default?

kaspth replied:
I must admit the setup line length got me to consolidate the tests. I do think you're right about keeping the old method, so we get a better test name, :+1:

senny commented on railties/test/application/test_runner_test.rb:
```diff
@@ -249,6 +252,73 @@ def test_generated_scaffold_works_with_rails_test
+      end
+    end
+
+    def test_run_multiple_files_and_run_one_file_by_line_only_runs_filter
```
Is this the behavior we want or a limitation by our current integration with minitest?

arthurnn replied:
this is a limitation of Minitest atm, that only allows you to pass one filter.
I have a semi-patch for Minitest to solve that. Until we send the PR and gets merged, thats the behaviour we can have.

arthurnn replied:
Sorry, what I said is not true.
We can implement a filter that accept files and lines, or just files. see that Minitest uses the filter ===. 
https://github.com/seattlerb/minitest/blob/master/lib/minitest.rb#L289

So our filter can hold a chain of files:line and we accept multiple results.

kaspth replied:
@arthurnn Right, I was thinking about adding a chain. Do we want to support passing several line filters?

senny replied:
@kaspth I'm not sure we need to support that. What got my attention is that these tests specify behavior that is caused by the current implementation not because we designed it like that. I think the behavior is surprising, you can pass arguments that are not at all considered. I think it would be better to either warn the user when that happens or get it working.

kaspth replied:
It was actually easier to just implement a `FilterChain` to let us run more line filters and augment Minitest's `-n` option.

kaspth replied:
But I just realized that doesn't solve your original concern, @senny. I'll try working on it some more.

senny commented:
@kaspth really liking this patch :yellow_heart: Can you post the help output when running with `-h`?

zenspider commented:
@senny it already does.

arthurnn commented on railties/lib/rails/test_unit/testing.rake:
```diff
@@ -1,12 +1,12 @@
+
+  ARGV.shift if ARGV[0] == "test"
+
+  Rails::TestRunner.run
```
why the runner dont receive the args?
I remember I initially had the `ARGV.shift` too, but I think, it didnt work when you do: `bundle exec rake test db:migrate` , dont fully remember why.

zenspider replied:
By the time a rake task is run, ARGV is out of the picture. That's how that task got run in the first place.

kaspth replied:
@zenspider Right, so as long as `rake` is the funnel people will have to keep using `TEST` and `TESTOPTS`.

kaspth replied:
I guess we could modify a `Rake::TestTask` to run something like `ruby -Ilib:test -r"rails/test_unit/minitest_plugin" whatever.rb`

arthurnn commented on railties/test/application/test_runner_test.rb:
```diff
@@ -284,6 +354,18 @@ def test_fixture
         RUBY
       end
 
+      def create_backtrace_test
```
the output of the tests that check backtrace look the same when backtrace is enable or disabled. Thats looks weird IMO. We should have different outputs I think.

kaspth replied:
Right, I can add a failure message to the `assert_kind_of` and match on that.

arthurnn commented on railties/test/application/test_runner_test.rb:
```diff
@@ -249,6 +252,73 @@ def test_generated_scaffold_works_with_rails_test
+      end
+    end
+
     private
```
should we also test the cases where tests fail, and the right output is there?

kaspth replied:
The way I see it, the happy path testing is to make sure we're correctly funneling to Minitest. Thus adding tests for more than our extensions would be testing Minitest.

arthurnn commented on activesupport/lib/active_support/test_case.rb:
```diff
@@ -45,6 +45,41 @@ def test_order
       end
+
+      def run(reporter, options = {})
+        if filter = line_filter_for(options[:patterns] || [])
```
why do we need this method here, and not in the Runner?
If we had this method in the runner, we could called it from the rails_minitest_plugin.
And tests that only extend Minitest::Test would also work.

kaspth replied:
We need to check which runnable implements a test - we don't have access to the runnable in `rails_minitest_plugin`.

Also adding it to Runner would make Active Support depend on Railties - this way stays clear of that.

arthurnn commented on railties/lib/rails/commands/test.rb:
```diff
@@ -2,4 +2,4 @@
 $: << File.expand_path("../../test", APP_PATH)
 
-Rails::TestRunner.run(ARGV)
+Minitest.autorun
```
why do we can the `autorun` here and not the runner?

senny replied:
regarding the naming discussion `TestRequirer` and `TestRunner`. I'd rather opt for something, which does not rely on `Minitest.autorun`.

@kaspth do you see any reasons not to go down the path of explicitly calling `Minitest.run` when we want it to run? Also clearing in between runs, making it possible to run different sets of tests in the same process?

arthurnn commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -1,14 +1,35 @@
+    ENV["RAILS_ENV"] = options[:environment] || "test"
+
+    options[:patterns] = OptionParser.new(options[:args]).order!
+    Rails::TestRunner.require_files options[:patterns]
```
I just realized we have this `require` call here, so do we need the `TestRunner.run` method at all?

arthurnn commented:
I like where this is going too.. Thanks @kaspth

senny commented:
@zenspider I meant to post the output here to GitHub.

zenspider commented:
"Runner" == manager class. That's what MT4 had and why I explicitly did a rewrite for MT5. Manager classes are bad design and should be avoided where possible. I explicitly designed test classes (runnables, really) to be able to run themselves on purpose. This allows for different types of runnables (eg Test vs Benchmark). Having a test class that knows how to run its own tests makes sense. It's what tests do: run, so why externalize the functionality of it?

Avoid a "Runner" type class. You'll be better off in the long run.

zenspider commented:
@kaspth Sorry, I missed your original question: you've removed 75 lines, of course I'm a fan.

senny commented:
@zenspider I can see your `Runner`-argument for Minitest itself but this runner does nothing else than requiring the files. Someone needs to require the tests before they can run themselves. I'd like to have an interface in Rails to run a set of files and folders. The rest boils down to wether `Runner` is a good name for that.

kaspth commented:
@zenspider Great to hear and, yes, I agree on removing the need for a runner and just use Minitest's `autorun` or `run` where appropriate.

@senny I'd be happy to go back to the `TestRequirer` name instead of `TestRunner`.

kaspth commented on activesupport/lib/active_support/testing/filter_chain.rb:
```diff
@@ -0,0 +1,51 @@
+
+      private
+        def derive_regexp(filter)
+          filter =~ %r%/(.*)/% ? Regexp.new($1) : filter
```
If we didn't duplicate some logic from Minitest here, passing a `-n /hello/` filter won't filter correctly.

kaspth commented:
@zenspider Writing the `FilterChain` is a bit cumbersome because we kind of have to hack Minitest. Would you consider letting others extend Minitest's filter in the same way reporters can?

kaspth commented:
@senny I forgot to post the help output. Here it is:

```
minitest options:
    -h, --help                                   Display this help.
    -s, --seed SEED                       Sets random seed
    -v, --verbose                             Verbose. Show progress processing files.
    -n, --name PATTERN                Filter run on /pattern/ or string.

Known extensions: rails
Usage: bin/rails test [options] [file or directory]
You can run a single test by appending the line number to filename:

bin/rails test test/models/user_test.rb:27
    -e, --environment [ENV]          Run tests in the ENV environment
    -b, --backtrace                         Show the complete backtrace
```

senny commented:
@kaspth I agree, the output is pretty dense at the moment. Also, the usage (`Usage: bin/rails test [options] [file or directory]`) needs adjusting that it's possible to pass multiple files or directories:

kaspth commented:
@senny I've tweaked the output to this:

```
minitest options:
    -h, --help                       Display this help.
    -s, --seed SEED                  Sets random seed
    -v, --verbose                    Verbose. Show progress processing files.
    -n, --name PATTERN               Filter run on /pattern/ or string.

Known extensions: rails

Usage: bin/rails test [options] [files or directories]
You can run a single test in file by appending the line number:

    bin/rails test test/models/user_test.rb:27

Rails options:
    -e, --environment [ENV]          Run tests in the ENV environment
    -b, --backtrace                  Show the complete backtrace
```

senny commented:
@kaspth looking better already. What do you think about a section similar to the one about the line number but describing to run multiple directories. Something along the lines of:

```
Usage: bin/rails test [options] [files or directories]

You can run a single test in file by appending the line number:

    bin/rails test test/models/user_test.rb:27

You can run multiple files and directories at the same time by passing multiple arguments:

    bin/rails test test/models test/controllers test/integration/login_test.rb

Rails options:
    -e, --environment [ENV]          Run tests in the ENV environment
    -b, --backtrace                  Show the complete backtrace
```

kaspth commented:
@senny Yeah, that's great. I've added it.

kaspth commented:
@senny @arthurnn Alright, I've added the ability to mix files and line filters as people please.

Something crazy like this should work:

```
test/controllers test/models/account_test.rb test/models/post_test.rb:4
```

Potentially we could output the failures on a single line, so you could copy that and run all the failing tests at once. But I don't know how useful that is.

I've also changed the code to not augment Minitest when there's no line filters.

However, when there's at least one line filter, we need to create filters for every file. If we didn't, running tests like:

```
test/models/account_test.rb test/models/post_test.rb:4
```

would only run the filter with the line on it. I feel like we should be able to avoid creating those extra filters, but I haven't been able to figure it out.

kaspth commented:
I really don't understand the two test failures in Railties. When I inspect the files required in those tests, they require the correct scaffolded files. And they require `test_helper` which requires `rails/test_help` which then calls `Minitest.autorun`.

Could ARGV being long gone in the rake test task be a factor in this?

senny commented on activesupport/lib/active_support/test_case.rb:
```diff
@@ -45,6 +46,15 @@ def test_order
       end
+
+      def run(reporter, options = {})
+        if options[:patterns] && options[:patterns].any? { |p| p.include?(':') }
```
Regarding the comment about [test-failures](https://github.com/rails/rails/pull/19571#issuecomment-94282124):

My guess is that `options[:patterns]` here is `["db:migrate", "test"]`. This does include a `:` resulting in a `CompositeFilter` being added.

kaspth commented:
@senny @arthurnn Alright, I think I got this now :smile:

I've moved away from `autorun` and fixed the errors in `rake_test.rb`. What should we do about `active_support/testing/autorun`? Can we outright remove it or do we have to deprecate?

kaspth commented:
Hm, now the generator tests are failing: https://travis-ci.org/rails/rails/jobs/61201646.

I'll look into them tomorrow (might be related to the lingering autorun @senny spotted).

senny commented:
@kaspth :+1: ping me and I'll take another look. Really happy to see this coming along. Great work :yellow_heart:

kaspth commented:
@senny I've been trying to debug why the generator tests are failing, but so far I've gone nowhere.

It seems the tests aren't run at all.

The generator generates an engine with a Rakefile with a standard rake test task. Which leaves me befuddled as to why the tests aren't run. Do you have any ideas?

senny commented:
@kaspth I can probably take a look later today.

senny commented:
@kaspth investigated a bit. The issue is the [removed autorun line in `test_help.rb`](https://github.com/rails/rails/pull/19571/files#diff-96b8e5694f7954d018d9958ef89f4049L6). As these tests are run using `Rake::TestTask`'s they expect that autorun will be required at some point.

zamith commented:
Hey guys,

First of all let me say that it is great to finally have a runner bundled with rails that can compete with the one from RSpec.

Also, as the maintainer of [m](https://github.com/qrush/m), I feel obligated to ask if there is something I can do to help out.

One thing I noticed is that this is very Minitest 5 specific, Rails 5 drops support for any previous version?

Thanks for the awesome job. :+1:

kaspth commented on railties/lib/rails/commands/test.rb:
```diff
@@ -1,5 +1,5 @@
 $: << File.expand_path("../../test", APP_PATH)
 
-Rails::TestRunner.run(ARGV)
+exit Minitest.run(ARGV)
```
I felt a bit weird adding `exit` here, but `autorun` also exits with what `Minitest.run` returns. Besides these tests failed without it: https://travis-ci.org/rails/rails/jobs/62532010#L819-L837.

arthurnn replied:
the exit make sense here, as we are not calling the autorun!

kaspth commented:
@senny @arthurnn Alright, we're green: https://travis-ci.org/rails/rails - it's kind of a bummer that GitHub doesn't show Travis runs for force pushes anymore :cry:

kaspth commented:
@zamith Just checking in the Gemfile we still accept versions below 5: https://github.com/rails/rails/blob/master/Gemfile#L65. But this pull request definitely relies heavily on the Minitest 5 extension system.

Personally, I think it would be great move to 5+ to move the community forward like we're doing with Ruby versions. What do you think, @arthurnn @senny?

zamith commented:
Yes, I agree. And I think that's the Rails philosophy, keep pushing the community. I'll keep `M` around for backwards compatibility, and non-Rails programs. It will also greatly diminish the complexity of the code base, because support all the versions of minitest and test::unit is a pain.

matthewd commented:
https://github.com/rails/rails/blob/master/activesupport/activesupport.gemspec#L26 -- we've been requiring minitest 5.1+ since Rails 4.1.

kaspth commented:
Right, Luís, that would be painful.

Great, Matthew! No worries then :smile:

> Den 14/05/2015 kl. 15.24 skrev Matthew Draper <notifications@github.com>:
> 
> https://github.com/rails/rails/blob/master/activesupport/activesupport.gemspec#L26 <https://github.com/rails/rails/blob/master/activesupport/activesupport.gemspec#L26> -- we've been requiring minitest 5.1+ since Rails 4.1.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/19571#issuecomment-102035924>.
> 

--
Kasper

kaspth commented on railties/lib/rails/test_help.rb:
```diff
@@ -3,13 +3,14 @@
 require 'action_dispatch/testing/integration'
 require 'rails/generators/test_case'
 
+require 'active_support/testing/autorun' unless $rails_test_runner
```
@senny I reintroduced the global variable to deal with the failures that ran with `rake test` and expected `test_help` to require `active_support/testing/autorun`.

arthurnn replied:
what we could do instead, is maybe set a class variable on the `Minitest` runner, and we could check that var?!?

arthurnn replied:
Something like ( on the minitest_plugin.rb )
```
module Minitest
 mattr_accessor :rails_runner
end
```

kaspth replied:
Yeah, that's better, :+1:. I settled on `run_with_rails_extension`. Thanks, @arthurnn.

senny commented:
@kaspth thanks a ton for the update. I'll take a look soon.

senny commented:
@kaspth code looks good.

I have a local application generated with `--dev`. When I try to run the tests I end up with the following exception:

```
$ bin/rails t
/Users/senny/Projects/rails/railties/lib/rails/test_unit/test_requirer.rb:9:in `require_files': undefined method `presence' for []:Array (NoMethodError)
	from /Users/senny/Projects/rails/railties/lib/rails/test_unit/minitest_plugin.rb:36:in `plugin_rails_init'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.6.1/lib/minitest.rb:74:in `block in init_plugins'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.6.1/lib/minitest.rb:72:in `each'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.6.1/lib/minitest.rb:72:in `init_plugins'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.6.1/lib/minitest.rb:123:in `run'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/test.rb:5:in `<top (required)>'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:128:in `require'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:128:in `require_command!'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:86:in `test'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:40:in `run_command!'
	from /Users/senny/Projects/rails/railties/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:8:in `require'
	from bin/rails:8:in `<main>'
```

Are we missing a `require "active_support"`?

kaspth commented:
@senny You're right, there was a `require 'active_support/core_ext/object/blank'` missing ('active_support` didn't require this, so I opted to require that file instead).

kaspth commented on railties/lib/rails/test_unit/reporter.rb:
```diff
@@ -15,8 +15,12 @@ def aggregated_results # :nodoc:
     end
+
+    def relative_path_for(file)
+      file.sub(/^#{Dir.pwd}\/?/, '')
```
I ran the runner in a `--dev` app I have and saw failure output like:

```
bin/rails test /Users/kasperhansen/Documents/code/collection_caching_test/test/integration/cache_invalidation_test.rb:26
```

To me the path doesn't really have to be absolute and it would be more readable to just have them output relatively. This change turns the above into:

```
bin/rails test test/integration/cache_invalidation_test.rb:26
```

What do you think, @senny @arthurnn?

gsamokovarov replied:
@kaspth Have you tried [`Pathname#relative_path_from`](http://ruby-doc.org/stdlib-2.1.0/libdoc/pathname/rdoc/Pathname.html#method-i-relative_path_from)? I think it can do it for you.

kaspth replied:
I experimented with it, but I couldn't get it to work like I wanted it to. And it was useless on absolute paths (which is what we're dealing with here).

gsamokovarov replied:
```ruby
>> p = Pathname("/Users/kasperhansen/Documents/code/collection_caching_test/test/integration/cache_invalidation_test.rb")
=> #<Pathname:/Users/kasperhansen/Documents/code/collection_caching_test/test/integration/cache_invalidation_test.rb>
>> p.relative_path_from(Pathname("/Users/kasperhansen/Documents/code/collection_caching_test"))
=> #<Pathname:test/integration/cache_invalidation_test.rb>'`
```

Doesn't that work for you? `Rails.root` is a `Pathname` already too.

kaspth replied:
Yup, that seems to work just fine, :+1:. Thanks, @gsamokovarov :smile:

kaspth replied:
Hm, that broke the build: https://travis-ci.org/rails/rails/jobs/63081213#L1323-L1359
I just reverted it.

gsamokovarov replied:
Yeah, that seems to be in the docs. Whatever works best then, thanks @kaspth!

kaspth replied:
No problem and thanks for the suggestion :smile:

> Den 19/05/2015 kl. 16.51 skrev Genadi Samokovarov <notifications@github.com>:
> 
> In railties/lib/rails/test_unit/reporter.rb <https://github.com/rails/rails/pull/19571#discussion_r30607211>:
> 
> >        end.join "\n"
> >      end
> > +
> > +    def relative_path_for(file)
> > +      file.sub(/^#{Dir.pwd}\/?/, '')
> Yeah, that seems to be in the docs. Whatever works best then, thanks @kaspth <https://github.com/kaspth>!
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/19571/files#r30607211>.
> 

--
Kasper

kaspth commented on railties/lib/rails/test_unit/reporter.rb:
```diff
@@ -15,8 +15,12 @@ def aggregated_results # :nodoc:
     end
+
+    def relative_path_for(file)
+      file.sub(/^#{Rails.root}\/?/, '')
```
So the [comment still stands](https://github.com/rails/rails/pull/19571#discussion_r30532322), I just used `Rails.root` instead.

kaspth commented on railties/test/application/test_test.rb:
```diff
@@ -64,8 +64,8 @@ def test_failure
-      assert_match %r{/app/test/unit/failing_test\.rb}, output
-      assert_match %r{/app/test/unit/failing_test\.rb:4}, output
+      assert_match %r{test/unit/failing_test\.rb}, output
+      assert_match %r{test/unit/failing_test\.rb:4}, output
```
Updated for relative paths.

kaspth commented:
And we're still green: https://travis-ci.org/rails/rails/builds/63059256 :smile:

zenspider commented:
I worked with @tenderlove and others to fix all the test order dependency bugs I could find (this is why https://github.com/seattlerb/debride was invented). The tests _should_ run in random order now and the dependency for `minitest < 5.3.4' _should_ be able to be lifted.

kaspth commented:
@zenspider Thanks for the suggestion. If the build is green with the latest changes, I'll push a commit that removes the restraint to see how we fare.

As far as I'm aware, there's been no effort to introduce more order dependence :wink:

kaspth commented:
Thanks @tgxworld, I should have a fix for what happens in Active Support here: https://github.com/rails/rails/pull/20267

arthurnn commented:
Ok, I did a few tests with this, and found a few problems:
- we need to add `require "active_support/core_ext/module/attribute_accessors"` to minitest_plugin.rb . as we use `mattr_accessor()`
- `ruby -Itest test/models/post_test.rb` is failing with ``test': test_the_truth is already defined in PostTest`

kaspth commented:
@arthurnn Thanks, I've added the missing require :smile:

However, I can't reproduce your test failure locally. Where are you running the command? 
I cd'ed into one of the apps generated by test runner test (added a simple Gemfile with my Rails clone, arel master and sqlite3), and ran `ruby -Itest test/models/post_test.rb` just fine.

I've added a test for this too: https://github.com/rails/rails/pull/19571/files#diff-f4d22edadb7ed6b4cc88e000d6c3ed69R254

arthurnn commented:
@kaspth I pushed my test app https://github.com/arthurnn/myapp . 
This is the full stacktrace of the error https://gist.github.com/a4189a0fb5bea964af56

kaspth commented:
@arthurnn Thanks, I've solved it now! :smile:

Basically when I set the `options[:patterns]` I expected `opts.order!` to return nil and not an empty array when there were no elements.
So `plugin_rails_init` would hit https://github.com/rails/rails/pull/19571/files#diff-ea6c92bf87b30213dbd5eafb8431976cR38, which erroneously would rerequire the files. I guess Rails' autoloading assumed this meant the file changed and began reloading it? Anyway it should work now :wink:

kaspth commented:
So my earlier solution broke a test running `bin/rails test` without arguments. Because `options[:patterns]` was now nil, it didn't require any files and no tests were run.

I've replaced this with a `Minitest.run_with_autorun` attribute. Basically when we require `test_help.rb` we can determine what we've been run with: https://github.com/rails/rails/pull/19571/files#diff-96b8e5694f7954d018d9958ef89f4049R12
If it wasn't the rails extension then it has to be autorun. Therefore `plugin_rails_init` won't attempt to require any files if we've been run with `autorun`. Because people should make sure all the files they need have been required if they aren't using the Rails test runner.

kaspth commented:
@arthurnn And we're green! https://travis-ci.org/rails/rails/builds/64804478 - the sqlite failure is the one master is plagued with :smile:

senny commented:
@kaspth I try to give this PR some testing tomorrow. Excited to see it getting close to the finish line.

senny commented:
@kaspth created a new app from this branch using `--dev`. Executing the runner without any tests fails with the following error:

```
$ bin/rails t                                                                                                                                                                                                        ~/Projects/playground/ruby/rails/test_runner
/Users/senny/Projects/rails/railties/lib/rails/test_unit/minitest_plugin.rb:40:in `plugin_rails_init': undefined method `backtrace_cleaner' for Rails:Module (NoMethodError)
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.7.0/lib/minitest.rb:74:in `block in init_plugins'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.7.0/lib/minitest.rb:72:in `each'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.7.0/lib/minitest.rb:72:in `init_plugins'
	from /Users/senny/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/minitest-5.7.0/lib/minitest.rb:123:in `run'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/test.rb:5:in `<top (required)>'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:128:in `require'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:128:in `require_command!'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:86:in `test'
	from /Users/senny/Projects/rails/railties/lib/rails/commands/commands_tasks.rb:40:in `run_command!'
	from /Users/senny/Projects/rails/railties/lib/rails/commands.rb:18:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

zenspider commented:
```
10003 % ag backtrace_ lib
lib/minitest.rb
28:  mc.send :attr_accessor, :backtrace_filter
226:    backtrace_filter.filter bt
776:  self.backtrace_filter = BacktraceFilter.new
```

kaspth commented:
@senny dang it, I'll take a look :smile:

@zenspider I'm not sure I follow?

senny commented:
@kaspth most likely the plugin currently depends on `test_help.rb` being required. Without tests, that's not the case.

kaspth commented:
@senny Right, I added a simple `respond_to`. I guess, it's not super clear why we have to check it. Let me know if you want me to change it :smile:

senny commented on railties/lib/rails/test_unit/minitest_plugin.rb:
```diff
@@ -1,14 +1,50 @@
+    Rails::TestRequirer.require_files options[:patterns] unless run_with_autorun
+
+    unless options[:full_backtrace] || ENV["BACKTRACE"]
+      Minitest.backtrace_filter = Rails.backtrace_cleaner if Rails.respond_to?(:backtrace_cleaner)
```
as you said the `respond_to?` is not very clear. Let's add a comment to describe why it's there.

senny commented:
@kaspth if you haven't already, can you add a test-case with a command that results in no runnables? Just to make sure we don't run into this problem again.

kaspth commented:
@senny Definitely. I'll add one now :+1:

kaspth commented:
@senny I can't figure out how to write a failing test for this. Rails is loaded at this point so it'll respond to `backtrace_cleaner` just fine.

```ruby

def test_run_app_without_tests
  Dir.chdir(app_path) do
    `rm -rf test`
    `bin/rails test`.tap do |output|
      assert_match '0 runs, 0 assertions', output
    end
  end
end
```

zenspider commented:
@kaspth That's because it is hard to follow an idiot making an idiotic comment. :) I've deleted my comment since I entirely missed the point. I'm blaming a lack of caffeine.

kaspth commented:
@zenspider haha, alright then :smile:

kaspth commented:
@senny I found out why the test doesn't reproduce the issue. Here's boot.rb from the generated app in the test:

```ruby
require '/Users/kasperhansen/Documents/code/rails/load_paths'
require 'rails/all'
```

And a real app:

```ruby
ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../../Gemfile', __FILE__)

require 'bundler/setup' # Set up gems listed in the Gemfile.
```

Not sure if we should just override that file or if there's a better way to simulate no runnables where Rails isn't loaded.

senny commented:
@kaspth let's start by overwriting the file. If someone has a more elegant idea we can change it later on.

/cc @rafaelfranca

rafaelfranca commented:
Yeah, I think overwriting is good.

kaspth commented:
@senny @rafaelfranca I've pushed the change overwriting the file.

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -229,24 +216,128 @@ def test_env
-          def test_env
-            puts Rails.env
+        class PostTest < ActiveSupport::TestCase
+          test 'declarative syntax works' do
```
This test would always pass because we were using `def test_the_truth`, which didn't cause the bug @arthurnn experienced with the RuntimeError saying "already defined in PostTest."

I'm not sure how to make this clearer.

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -229,24 +216,128 @@ def test_env
 
-      assert_match env, run_test_command("-e #{env} test/unit/env_test.rb")
+      Dir.chdir(app_path) do
+        `ruby -Itest test/models/post_test.rb`.tap do |output|
```
Also if we ever regress here `RuntimeError` just goes straight through here. I've tried catching it, but I guess the fact that we're in another process(?) throws us off. The tests will be marked as a failure, but we'll never hit the assertions below.

senny commented:
Just gave the latest version of this PR a spin and everything worked. (Even did some spring testing). Help text is descriptive and the new integration is superb. Thank you @kaspth for your work! :yellow_heart: :yellow_heart: :yellow_heart:

kaspth commented:
Woot! So glad to have this merged :heart::yellow_heart::purple_heart::green_heart::blue_heart:

arthurnn commented:
NICE JOB!!!!

deivid-rodriguez commented:
@kaspth Thanks a lot for this, it's a great improvement! :smile:
